### DISCUSSION TITLE
Trouble using packer_plugins global

### DISCUSSION THREAD

**Discussion by strayer** [wbthomason/packer.nvim#196] [Q&A] [answered]
I'm trying to utilize the new `packer_plugins` global to configure stuff in lua files only for installed plugins.

For example, I only want to configure lightlines nvim-coc support if coc is actually installed:

```lua
-- statusline.lua
[...]

if packer_plugins["coc-nvim"] and packer_plugins["coc-nvim"].loaded then
  vim.fn['lightline#coc#register']()
end
```

My config is structured into multiple files, so my init.lua (technically a start.lua in lua/ required in init.vim right now until I can move everything over to an actual init.lua) does this:

```lua
require('plugins') -- contents as described in this projects README.md
require('statusline')
```

This causes nvim to throw this error when starting up:

> Fehler beim Ausführen von "/Users/strayer/.config/nvim/init.vim":
> Zeile    4:
> E5108: Error executing lua /Users/strayer/.config/nvim/lua/statusline.lua:12: attempt to index global 'packer_plugins' (a nil value)

I'm new to lua and probably doing something very stupid here, but as far as I understand global variables should be available in another required file too. 🤷‍♂️

**Comment by wbthomason** [comment, marked as answer]
I think this is probably an issue regarding startup sourcing order. `packer_compiled` is in `plugin/` (by default), which gets sourced **after** your `init.vim`. So, if you're doing these checks in your `init.vim`, `packer_plugins` doesn't exist yet. (you can see more on startup order with `:help startup`)

There are maybe two ways to deal with this. Option one is to move this code to a file in `plugin/` that gets sourced after `packer_compiled`. Option two is to configure `packer` to put the compiled file somewhere else and manually source it in your `init.vim` before you run this code.

**Reply by strayer** [reply]
Ah, that makes sense. Regarding the first suggestion: This complicates my setup a bit because I set stuff in the lua files that need to be set in init.* (e.g. termguicolors) that will cause errors when I only require in a vim file in plugins/.

Regarding configuring packer to put the compiled file somewhere else: Is there a caveat to this except having to explicitly load the compiled vim file in init.lua/vim?

I think this may confuse a lot of people. I mainly use this to only run some configurations if a plugin is actually installed. I share my neovim config on multiple machines, but don't want to install coc on all of them, for example.

**Reply by strayer** [reply]
I solved this for now by moving the packer_compiled.vim to $HOME/.config/nvim/packer_compiled.vim and sourcing it myself:

```vimscript
" init.vim
lua require('plugins')
source $HOME/.config/nvim/packer_compiled.vim
lua require('start')
```

```lua
-- lua/plugins.lua
vim.cmd [[packadd packer.nvim]]

util = require('packer/util')

return require('packer').startup({function()
  -- Packer can manage itself as an optional plugin
  use {'wbthomason/packer.nvim', opt = true}

  -- ...
end, config = {compile_path = util.join_paths(vim.fn.stdpath('config'), 'packer_compiled.vim')}});
```

**Reply by wbthomason** [reply]
> Regarding configuring packer to put the compiled file somewhere else: Is there a caveat to this except having to explicitly load the compiled vim file in init.lua/vim?

There should not be - if there is, it's a bug.

> I think this may confuse a lot of people. I mainly use this to only run some configurations if a plugin is actually installed. I share my neovim config on multiple machines, but don't want to install coc on all of them, for example.

May I ask how you conditionally install plugins? It seems like keeping your configuration in a `config` or `setup` key and using either the `disabled` key or `cond` key to control installation might be a simpler/cleaner setup.

**Reply by strayer** [reply]
Before switching to packer I simply used variables and toggled some plugins based on my hostname.

```
" init.vim

let s:hostname = trim(system('hostname'))
let s:should_install_coc = s:hostname == 'khitomer'

" ...

" COC
if s:should_install_coc
  Plug 'neoclide/coc.nvim', {'branch': 'release'}
  " ...
endif

" ...

if s:should_install_coc
  " coc-specific functions/configurations/etc here
endif
```

This is essentially what you are suggesting with `disabled`/`cond`, right? I have to admit that I don't yet conditionally install coc with packer because I do the migration from vimscript to lua on the machine where all packages are installed anyway - I guess I would have used `cond` and check for the correct hostname(s) there.

Ideally I would like to base all configuration and such simply on the condition that the plugin is there and enabled. I think this keeps the configuration cleaner because I don't need custom conditions per package but can simply assume it needs to be configured if it is there.

**Reply by wbthomason** [reply]
If I understand what you want, I think you should be able to write something like:
```lua
local enable_coc = vim.fn.trim(vim.fn.system('hostname')) == 'khitomer'
use {'neoclide/coc.nvim', branch = 'release', disabled = not enable_coc, config = YourCocConfigFunction}
```

More generally, the `config` key is only used if a plugin is present, so whether you're determining whether to install a plugin via `disabled` or some other mechanism, simply putting your plugin-specific configuration in a `config`/`setup` key should handle conditionally including those functions, etc. for you.

**Reply by strayer** [reply]
Ohhh, I completely overlooked that. This is much better than what I was doing before and should cover all things I intend to do with this. And as a bonus also doesn't require me explicitly loading the compiled packer code – awesome.

Thanks a lot for pointing this out and helping me along the way!

**Reply by wbthomason** [reply]
Cool, I'm glad things are working out better!
